### ISSUE TITLE
Include OpenH264 library when compiling

### ISSUE THREAD

**Issue by powerqian** [BelledonneCommunications/linphone-android#108]
Dear Linphone Android developers,

I understand that user can download OpenH264 codec library in the app, but I'm wondering is it possible to directly include it when compiling liblinphone SDK? It may have the following advantages.

- User can use H264 codec out of box without extra waiting time or configuration.
- Based on the documentation, the downloaded OpenH264 library only works for ARM architecture. However OpenH264 can be built for `arm`, `arm64`, `x86` and `x86_64`. I guess we can manually build the OpenH264 library, include them somewhere when compiling liblinphone SDK, and enable the H264 codec for all those architectures?

I've tried to read the source code of the OpenH264 downloader in the liblinphone, and manually copy/paste the libopenh264 to the destination where it downloads the library when the app is launched. In that way the downloader will treat it as downloaded, and enable the H264 codec using the OpenH264 directly. However it is not robust and still only works for `arm` architecture I think.

Is it possible to improve this kind of behaviour?

**Comment by DatBewar** [comment]
Hello,
You can read our wiki concerning OpenH264 downloader.
And if you want to embeb directly OpenH264 you can run prepare.py:
./prepare.py -DENABLE_NON_FREE_CODECS=1 -DENABLE_OPENH264=1 -DENABLE_EMBEDDED_OPENH264=1

**Comment by powerqian** [comment]
Hi @DatBewar 

Thank you for your reply.

After checking log and doing other experiment, eventually I found out that the key is to remove the following line in all the gradle script
```
excludePackage.add('**/libopenh264**')
```
This line will exclude the `libopenh264.so` from the final build. Commenting that line, adding the following line to the static load in `LinphoneCoreFactoryImpl`, remove all code that related to OpenH264Downloader, and use H264Helper to set the mode to `MODE_AUDO` will do the trick.
```
System.loadLibrary("openh264");
```

By the way, the `msopenh264` submodule in current `master` [3e0b1dc9711114fc0229e354e025656471ff0693] and `beta_android_3.2.8` [97f8c0c98e42f68f49a755bf1667abaeea92cab6] needs to be updated. Otherwise it cannot be loaded due to some dependency not found issue, which is fixed in 18b41c2569ea76b8eb1df9f85a32f16b3d001cc0 of `msopenh264`

**Comment by ArtemDivin** [comment]
@powerqian hi, i try to resolve the same problem, but couldn't build correct sdk, could you share your build sdk please?
